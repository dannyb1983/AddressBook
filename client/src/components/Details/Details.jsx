import React, { Fragment, useState, useEffect } from "react";
import { Form, Col, Row, Container, Button } from "react-bootstrap";
import { useHistory } from "react-router";

const Details = ({
  people,
  contactToEdit,
  deleteContact,
  sortPeople,
  sortType,
  sameFirstLastNameCheck,
}) => {
  const [emptyName, setEmptyName] = useState(null);
  const history = useHistory();

  useEffect(() => {
    if (!contactToEdit || !people) {
      history.push("/");
    }
  }, []);
  // Back button
  const goBack = () => {
    history.push("/");
  };
  // Save the user that's being edited
  const saveEdit = (e) => {
    e.preventDefault();
    // Create a new contact to replace the old one, retaining the uuid/ photo from original contact
    const newContact = {};
    newContact.login = { uuid: contactToEdit.login.uuid };
    newContact.name = { first: e.target[0].value, last: e.target[1].value };
    newContact.email = e.target[2].value;
    newContact.phone = e.target[3].value;
    newContact.picture = { thumbnail: contactToEdit.picture.thumbnail };
    // If the there are no changes to made to the contact
    if (
      newContact.name.first === contactToEdit.name.first &&
      newContact.name.last === contactToEdit.name.last &&
      newContact.email === contactToEdit.email &&
      newContact.phone === contactToEdit.phone
    ) {
      history.push("/");
      return;
    }
    // If first or last name was made blank
    if (newContact.name.first.length < 1 || newContact.name.last.length < 1) {
      setEmptyName(true);
      return;
    } // If there is another contact with the same first & last name
    if (sameFirstLastNameCheck(newContact) === true) {
      let sameUserConfirmation = window.confirm(
        "A contact with this name already exists. Are you sure you want to continue?"
      );
      if (!sameUserConfirmation) {
        return;
      } else {
        let deletedContacts = deleteContact(contactToEdit);
        sortPeople([...deletedContacts, newContact], sortType);
        history.push("/");
      }
    } else {
      let deletedContacts = deleteContact(contactToEdit);
      sortPeople([...deletedContacts, newContact], sortType);
      history.push("/");
    }
  };
  // Delete use confirmation
  const deletePermission = () => {
    let userConfirmation = window.confirm(
      "Are you sure you want to delete this Contact?"
    );
    if (userConfirmation) {
      deleteContact(contactToEdit);
      history.push("/");
    }
  };


  return (
    <Fragment>
      {contactToEdit && (
        <>
          <Container className="p-3">
            <Row>
              <Col md={12} className="text-left py-5 mt-3 ml-0 pl-0">
                <h1>Contact Details</h1>
                <div style={{ paddingLeft: "30px" }}>
                  <Button variant="secondary" onClick={goBack}>
                    <i className="fas fa-arrow-left mr-2"></i>back
                  </Button>
                </div>
              </Col>
            </Row>
            <Fragment>
              <Form onSubmit={saveEdit}>
                <Form.Group
                  as={Row}
                  className="mb-3"
                  controlId="formPlaintextEmail"
                >
                  <Form.Label column sm="2">
                    First Name
                  </Form.Label>
                  <Col sm="10">
                    <Form.Control
                      defaultValue={contactToEdit.name.first}
                      type="text"
                    />
                  </Col>
                </Form.Group>
                <Form.Group
                  as={Row}
                  className="mb-3"
                  controlId="formPlaintextEmail"
                >
                  <Form.Label column sm="2">
                    Last Name
                  </Form.Label>
                  <Col sm="10">
                    <Form.Control
                      defaultValue={contactToEdit.name.last}
                      type="text"
                    />
                  </Col>
                </Form.Group>
                <Form.Group
                  as={Row}
                  className="mb-3"
                  controlId="formPlaintextEmail"
                >
                  <Form.Label column sm="2">
                    Email
                  </Form.Label>
                  <Col sm="10">
                    <Form.Control
                      defaultValue={contactToEdit.email}
                      type="email"
                    />
                  </Col>
                </Form.Group>
                <Form.Group
                  as={Row}
                  className="mb-3"
                  controlId="formPlaintextEmail"
                >
                  <Form.Label column sm="2">
                    Phone
                  </Form.Label>
                  <Col sm="10">
                    <Form.Control
                      defaultValue={contactToEdit.phone}
                      type="input"
                    />
                  </Col>
                </Form.Group>
                <div style={{ padding: "10px", textAlign: "center" }}>
                  <Button
                    onClick={deletePermission}
                    variant="primary"
                    type="button"
                  >
                    Delete Contact
                  </Button>
                  <div
                    style={{
                      width: "20px",
                      height: "auto",
                      display: "inline-block",
                    }}
                  />
                  <Button variant="primary" type="submit">
                    Save Contact
                  </Button>
                </div>
              </Form>
              {emptyName ? (
                <h5
                  style={{ color: "red", padding: "20px", textAlign: "center" }}
                >
                  **Contacts must contain a first AND last name
                </h5>
              ) : null}
            </Fragment>
          </Container>
        </>
      )}
    </Fragment>
  );
};
export default Details;
