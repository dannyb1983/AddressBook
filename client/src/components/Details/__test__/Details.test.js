//modules
import React from 'react';
import { render, screen, cleanup} from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { contact, contacts } from '../../../../../test/jest/__mocks__/testData';
//components
import Details from '../Details';

describe('Details (Unit)', () => {
  afterEach(() => cleanup());
  const renderDetails = () => render(<MemoryRouter initialEntries={['/details']}><Details emptyName={true} people={contacts} contactToEdit={contact} deleteContact={()=>people.pop()} sortPeople={()=>people} sortType={'first'} sameFirstLastNameCheck={false} /></MemoryRouter>)
  const renderEmptyDetails = () => render(<MemoryRouter><Details emptyName={true} /></MemoryRouter>)

  describe('render()', () => {
    const { asFragment, container, } = renderDetails();

    it('render should match snapshot', () => {
        renderDetails()
      expect(asFragment().firstChild).toMatchSnapshot();
    });
    it('should render h1', () => {
      renderDetails()
      expect(screen.getByRole('heading',{name:'Contact Details'})).toBeInTheDocument();
    });
    it('Should not render a warning message ', () => {
      renderDetails()
      const nofirstOrLastName = screen.queryByText('**Contacts must contain a first AND last name')
      expect(nofirstOrLastName).not.toBeInTheDocument()
    });
    
  });
});
