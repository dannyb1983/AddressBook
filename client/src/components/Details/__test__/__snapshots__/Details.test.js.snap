// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details (Unit) render() render should match snapshot 1`] = `
<div
  class="p-3 container"
>
  <div
    class="row"
  >
    <div
      class="text-left py-5 mt-3 ml-0 pl-0 col-md-12"
    >
      <h1>
        Contact Details
      </h1>
      <div
        style="padding-left: 30px;"
      >
        <button
          class="btn btn-secondary"
          type="button"
        >
          <i
            class="fas fa-arrow-left mr-2"
          />
          back
        </button>
      </div>
    </div>
  </div>
  <form
    class=""
  >
    <div
      class="mb-3 form-group row"
    >
      <label
        class="form-label col-form-label col-sm-2"
        for="formPlaintextEmail"
      >
        First Name
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="formPlaintextEmail"
          type="text"
          value="John"
        />
      </div>
    </div>
    <div
      class="mb-3 form-group row"
    >
      <label
        class="form-label col-form-label col-sm-2"
        for="formPlaintextEmail"
      >
        Last Name
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="formPlaintextEmail"
          type="text"
          value="Smith"
        />
      </div>
    </div>
    <div
      class="mb-3 form-group row"
    >
      <label
        class="form-label col-form-label col-sm-2"
        for="formPlaintextEmail"
      >
        Email
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="formPlaintextEmail"
          type="email"
          value="brett.bishop@example.com"
        />
      </div>
    </div>
    <div
      class="mb-3 form-group row"
    >
      <label
        class="form-label col-form-label col-sm-2"
        for="formPlaintextEmail"
      >
        Phone
      </label>
      <div
        class="col-sm-10"
      >
        <input
          class="form-control"
          id="formPlaintextEmail"
          type="input"
          value="(326)-191-0475"
        />
      </div>
    </div>
    <div
      style="padding: 10px; text-align: center;"
    >
      <button
        class="btn btn-primary"
        type="button"
      >
        Delete Contact
      </button>
      <div
        style="width: 20px; height: auto; display: inline-block;"
      />
      <button
        class="btn btn-primary"
        type="submit"
      >
        Save Contact
      </button>
    </div>
  </form>
</div>
`;
